protocol/daemons/server/types/health_monitor.go

```
@@ -0,0 +1,293 @@
+	logger             log.Logger
+	startupGracePeriod time.Duration
+	pollingFrequency   time.Duration
+}
+
+// NewHealthMonitor creates a new health monitor.
+func NewHealthMonitor(
+	startupGracePeriod time.Duration,
+	pollingFrequency time.Duration,
+	logger log.Logger,
+) *HealthMonitor {
+	return &HealthMonitor{
+		serviceToHealthChecker: make(map[string]*healthChecker),
+		logger:                 logger.With(cosmoslog.ModuleKey, "health-monitor"),
+		startupGracePeriod:     startupGracePeriod,
+		pollingFrequency:       pollingFrequency,
+	}
+}
+
+func (hm *HealthMonitor) DisableForTesting() {
+	hm.lock.Lock()
+	defer hm.lock.Unlock()
+
+	hm.disabled = true
+}
+
+// RegisterServiceWithCallback registers a HealthCheckable with the health monitor. If the service
+// stays unhealthy every time it is polled during the maximum acceptable unhealthy duration, the monitor will
+// execute the callback function.
+// This method is synchronized. The method returns an error if the service was already registered or the
+// monitor has already been stopped.
+func (hm *HealthMonitor) RegisterServiceWithCallback(
+	hc types.HealthCheckable,
+	maximumAcceptableUnhealthyDuration time.Duration,
+	callback func(error),
+) error {
+	hm.lock.Lock()
+	defer hm.lock.Unlock()
+
+	if maximumAcceptableUnhealthyDuration <= 0 {
+		return fmt.Errorf(
+			"health check registration failure for service %v: "+
+				"maximum acceptable unhealthy duration %v must be positive",
+			hc.ServiceName(),
+			maximumAcceptableUnhealthyDuration,
+		)
+	}
+
+	// Don't register daemon services if the monitor has been disabled.
+	if hm.disabled {
+		return nil
+	}
+
+	// Don't register additional daemon services if the monitor has already been stopped.
+	// This could be a concern for short-running integration test cases, where the network
+	// stops before all daemon services have been registered.
+	if hm.stopped {
+		return fmt.Errorf(
```

I think we should stop the `HealthCheckable` and return `nil`.

This will prevent tests that register the daemons from panicking due to the failed registration if the timing of the test caused it to shutdown first and will also propagate the shutdown to the `HealthCheckable`.

Please also update the comment of `RegisterService`/`RegisterServiceWithCallback`

---

Not all daemons are stoppable at this time, but I added some code that conditionally stops daemons with a `Stop` method.

---

Is there a reason why we still want to return an error here?

---

I thought of it as informative, but yes, it's not truly an error condition.